### PR TITLE
PIM-3860: Fix filters, setters and denormalizers for reference data

### PR DESCRIPTION
**Pull request by jjanvier**
| Q                    | A
| -------------------- | ---
| Bug fix?             | N
| New feature?         | N
| BC breaks?           | N
| CI currently passes? | Y
| Tests pass?          | Y (except 2 because a merge of master is required)
| Scenarios pass?      | running
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   |
| Fixed tickets        |
| DB schema updated?   |
| Migration script?    |
| Doc PR               |

### CODE REVIEW

File: spec/Pim/Bundle/ReferenceDataBundle/Doctrine/ReferenceDataRepositoryResolverSpec.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        $this->shouldHaveType('Pim\Bundle\ReferenceDataBundle\Doctrine\ReferenceDataRepositoryResolver');
+    }
+
+    function it_resolves_the_repository_of_a_reference_data(
+        $configurationRegistry,
+        $doctrine,
+        ConfigurationInterface $configuration,
+        ObjectRepository $repository
+
```
Comment (skeleton) [inline spec/Pim/Bundle/ReferenceDataBundle/Doctrine/ReferenceDataRepositoryResolverSpec.php]: blank line

File: spec/Pim/Bundle/ReferenceDataBundle/Doctrine/ReferenceDataRepositoryResolverSpec.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    ) {
+        $configuration->getClass()->willReturn('Acme\Bundle\AppBundle\Entity\Color');
+        $configurationRegistry->get('colors')->willReturn($configuration);
+        $doctrine->getRepository('Acme\Bundle\AppBundle\Entity\Color')->willReturn($repository);
+
+
```
Comment (skeleton) [inline spec/Pim/Bundle/ReferenceDataBundle/Doctrine/ReferenceDataRepositoryResolverSpec.php]: double blank line

File: spec/Pim/Bundle/ReferenceDataBundle/Doctrine/ReferenceDataRepositoryResolverSpec.php (hunk 1 above)
Comment (jjanvier) [inline spec/Pim/Bundle/ReferenceDataBundle/Doctrine/ReferenceDataRepositoryResolverSpec.php]: done

File: src/Pim/Bundle/CatalogBundle/Updater/Setter/ReferenceDataCollectionSetter.php [hunk 3]
```diff
@@ -51,9 +50,28 @@ public function setAttributeData(
         $this->checkLocaleAndScope($attribute, $options['locale'], $options['scope'], 'reference data collection');
         $this->checkData($attribute, $data);
 
-        $data = $this->referenceDataDenormalizer->denormalize($data, '', null, ['attribute' => $attribute]);
+        $referenceDataCollection = [];
+        foreach ($data as $referenceDataCode) {
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Updater/Setter/ReferenceDataCollectionSetter.php]: remove blank line inside foreach

File: src/Pim/Bundle/CatalogBundle/Updater/Setter/ReferenceDataCollectionSetter.php [hunk 4]
```diff
@@ -51,9 +50,28 @@ public function setAttributeData(
...
 
-        $this->setReferenceDataCollection($attribute, $product, $data, $options['locale'], $options['scope']);
+            $repository = $this->repositoryResolver->resolve($attribute->getReferenceDataName());
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Updater/Setter/ReferenceDataCollectionSetter.php]: you don't need to find repository in each item of foreach

File: src/Pim/Bundle/CatalogBundle/Updater/Setter/ReferenceDataCollectionSetter.php [hunk 5]
```diff
@@ -51,9 +50,28 @@ public function setAttributeData(
...
+                    'setter',
+                    'reference data collection',
+                    $referenceDataCode
+                );
+            }
+
+            $referenceDataCollection[] = $referenceData;
+        }
+
+
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Updater/Setter/ReferenceDataCollectionSetter.php]: double blank line

File: src/Pim/Bundle/ReferenceDataBundle/Doctrine/ReferenceDataIdResolver.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    public function __construct(ReferenceDataRepositoryResolver $repositoryResolver)
+    {
+        $this->repositoryResolver = $repositoryResolver;
+    }
+
+    /**
+     * Resolve reference data IDs from codes
+     *
+     * @param string $referenceData reference data name
+     * @param mixed  $codes a code or an array of code
```
Comment (skeleton) [inline src/Pim/Bundle/ReferenceDataBundle/Doctrine/ReferenceDataIdResolver.php]: align phpdoc

File: src/Pim/Bundle/ReferenceDataBundle/Doctrine/ReferenceDataRepositoryResolver.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ */
+class ReferenceDataRepositoryResolver
+{
+    /** @var ConfigurationRegistryInterface */
+    protected $configurationRegistry;
+
+    /** @var RegistryInterface */
+    protected $doctrineRegistry;
+
+    public function __construct(
```
Comment (skeleton) [inline src/Pim/Bundle/ReferenceDataBundle/Doctrine/ReferenceDataRepositoryResolver.php]: phpdoc

File: src/Pim/Bundle/ReferenceDataBundle/Doctrine/ReferenceDataRepositoryResolver.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        ConfigurationRegistryInterface $configurationRegistry,
+        RegistryInterface $doctrineRegistry
+    ) {
+        $this->configurationRegistry = $configurationRegistry;
+        $this->doctrineRegistry      = $doctrineRegistry;
+    }
+
+    /**
+     * @param $referenceData
```
Comment (skeleton) [inline src/Pim/Bundle/ReferenceDataBundle/Doctrine/ReferenceDataRepositoryResolver.php]: missing type of $referenceData

File: src/Pim/Component/ReferenceData/Updater/Copier/ReferenceDataCollectionAttributeCopier.php [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+use Pim\Bundle\CatalogBundle\Model\AttributeInterface;
+use Pim\Bundle\CatalogBundle\Model\ProductInterface;
+use Pim\Bundle\CatalogBundle\Model\ProductValueInterface;
+use Pim\Bundle\CatalogBundle\Updater\Copier\AbstractAttributeCopier;
+use Pim\Bundle\CatalogBundle\Validator\AttributeValidatorHelper;
+use Pim\Component\ReferenceData\MethodNameGuesser;
+
+/**
+ * Copy a reference data collection value attribute in other reference data collection value attribute
+ * @author    Julien Janvier <jjanvier@akeneo.com>
```
Comment (skeleton) [inline src/Pim/Component/ReferenceData/Updater/Copier/ReferenceDataCollectionAttributeCopier.php]: add blank line

**Pull request merged**